### PR TITLE
[quant] Add int8 per token dynamic quant + int4 per group quant for ExecuTorch

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #102

Summary:
att

Adding this for accuracy evaluation, we also added this in executorch repo and we'll dedup later

Test Plan:

quantization:
```
python quantize.py --checkpoint_path checkpoints/$MODEL_REPO/model.pth --mode 8da4w-gptq --calibration_tasks wikitext --calibration_limit 5
```
this finished in 20+ min in my machine
if you change calibration_limit to 1, then it can be finished in 10+ min, but expect worse quality since we do less calibration (use this for debugging a new quantization experiment)

evaluation:
```
python eval.py --checkpoint_path checkpoints/$MODEL_REPO/model_8da4w-gptq.g32.pth --tasks wikitext
```
This should be fast, the result I'm getting is:
```
wikitext: {'word_perplexity,none': 10.15655335078972, 'word_perplexity_stderr,none': 'N/A', 'byte_perplexity,none': 1.5726497149737177, 'byte_perplexity_stderr,none': 'N/A', 'bits_per_byte,none': 0.6531973670369153, 'bits_per_byte_stderr,none': 'N/A', 'alias': 'wikitext'}
```

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: et_quantize.py [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+    """8 bit per token dynamic quantization for activation and 4 bit per channel group quantization
+    for weight
+    """
+    x = per_token_dynamic_quant(x)
+    origin_x_size = x.size()
+    x = x.reshape(-1, origin_x_size[-1])
+    scales_and_zeros = scales_and_zeros.to(torch.float)
+    w_dq = group_dequantize_tensor_symmetric(weight_int8, scales_and_zeros, groupsize, precision)
+    x = x.to(precision)
+    c = torch.ops.aten.linear(x, w_dq)
```
Comment (kimishpatel) [inline et_quantize.py:470]: For fp16, we will ahve to make some changes here that may result in the need of a different pattern in xnnpack quantizer

File: quantize.py [hunk 2]
```diff
@@ -516,6 +518,193 @@ def forward(self, input: torch.Tensor) -> torch.Tensor:
...
+        input = input.to(self.precision)
+        input = F.pad(input, pad=(0, self.in_features - self.origin_in_features))
+        return linear_forward_8da4w(
+            input,
+            self.weight, self.scales_and_zeros, self.out_features, self.groupsize, self.precision
+        )
+
+
+class Int8DynActInt4WeightGPTQQuantHandler(GPTQQuantHandler):
+    def __init__(self, mod, groupsize=128, inner_k_tiles=8, padding_allowed=True):
```
Comment (kimishpatel) [inline quantize.py:577]: what is the significance of inner_k_tiles

Comment (HDCharles) [reply quantize.py:577]: kernel hyperparameter

File: model.py [hunk 3]
```diff
@@ -72,10 +72,8 @@ def update(self, input_pos, k_val, v_val):
         # input_pos: [S], k_val: [B, H, S, D]
         assert input_pos.shape[0] == k_val.shape[2]
 
-        k_out = self.k_cache
-        v_out = self.v_cache
-        k_out[:, :, input_pos] = k_val
-        v_out[:, :, input_pos] = v_val
+        k_out = torch.ops.aten.index_put_(self.k_cache, [None, None, input_pos], k_val)
```
Comment (cpuhrsch) [inline model.py:75]: Why is this change needed?

File: model.py [hunk 4]
```diff
@@ -223,7 +221,7 @@ def precompute_freqs_cis(
     freqs = torch.outer(t, freqs)
     freqs_cis = torch.polar(torch.ones_like(freqs), freqs)
     cache = torch.stack([freqs_cis.real, freqs_cis.imag], dim=-1)
-    return cache.to(dtype=torch.bfloat16)
+    return cache.to(dtype=torch.float16)
```
Comment (cpuhrsch) [inline model.py:224]: I assume this needs to become a parameter

File: model.py (hunk 3 above)
Comment (jerryzh168) [reply model.py:75]: dynamo.export does not support capturing mutation (the mutated tensor will be written to a new tensor) I think, this is added to work around that, @HDCharles have more context on this

File: model.py (hunk 4 above)
Comment (jerryzh168) [reply model.py:224]: yeah, will make this a parameter before landing

**Comment by jerryzh168** [comment]
we're going to add this to torchao instead
